node/bft-consensus/src/lib.rs

```diff
@@ -186,3 +192,37 @@ fn read_authority_keypair_from_file<P: AsRef<std::path::Path>>(path: P) -> anyho
     let contents = std::fs::read_to_string(path)?;
     BLS12381KeyPair::decode_base64(contents.as_str().trim()).map_err(|e| anyhow!(e))
 }
+
+#[derive(Clone)]
+struct TransactionValidator<N: Network, C: ConsensusStorage<N>>(AleoConsensus<N, C>);
+
+impl<N: Network, C: ConsensusStorage<N>> narwhal_worker::TransactionValidator for TransactionValidator<N, C> {
+    type Error = anyhow::Error;
+
+    /// Determines if a transaction valid for the worker to consider putting in a batch
+    fn validate(&self, transaction: &[u8]) -> Result<(), Self::Error> {
+        let bytes = BytesMut::from(transaction);
+        let message = Message::deserialize(bytes)?;
+
+        let unconfirmed_transaction = if let Message::UnconfirmedTransaction(unconfirmed_transaction) = message {
+            unconfirmed_transaction
+        } else {
+            bail!("[UnconfirmedTransaction] Expected Message::UnconfirmedTransaction, got {:?}", message.name());
+        };
+
+        let transaction = match unconfirmed_transaction.transaction.deserialize_blocking() {
+            Ok(transaction) => transaction,
+            Err(error) => bail!("[UnconfirmedTransaction] {error}"),
+        };
+        self.0.add_unconfirmed_transaction(transaction)?;
```

Comment:
Is there a chance this function might be called more than once for same transaction and idempotency is assumed?

Reply:
this is currently not called anywhere by the validator, but if it were to be called twice, it wouldn't do anything (the duplicate would be ignored)